File: smart-hub/src/main/scala/io/github/tassiLuca/hub/adapters/ui/DashboardUI.scala
package io.github.tassiLuca.hub.adapters.ui

import io.github.tassiLuca.utils.ScalaSwingFacade.{*, given}

import java.awt.{BorderLayout, Font, GridLayout}
import javax.swing.*
import javax.swing.table.DefaultTableModel

class DashboardUI extends JFrame:
  private val biggerFont = Font("Arial", Font.BOLD, 20)

  setDefaultCloseOperation(WindowConstants.EXIT_ON_CLOSE)
  setTitle("Smart hub dashboard")
  setSize(915, 460)
  setLocationRelativeTo(null)
  setLayout(GridLayout(1, 2))
  val temperatureLabel: JLabel = JLabel("--")
  temperatureLabel.setFont(biggerFont)
  val heaterLabel: JLabel = JLabel("--")
  heaterLabel.setFont(biggerFont)
  val luminosityModel: DefaultListModel[String] = DefaultListModel[String]()
  private val luminosityEntries = JList(luminosityModel)
  private val infos = createPanel(
    JSeparator(),
    createPanel(JLabel("Current Temperature:"), temperatureLabel),
    createPanel(JLabel("Heater Status:"), heaterLabel),
    createPanel(JLabel("Current Luminosity:"), luminosityEntries),
  )
  infos.setLayout(BoxLayout(infos, BoxLayout.PAGE_AXIS))
  val alertsModel: DefaultTableModel = DefaultTableModel(Array[AnyRef]("Alert Messages"), 0)
  private val alertsLog: JTable = JTable(alertsModel)
  private val alerts = createPanel(JScrollPane(alertsLog))
  alerts.setLayout(BoxLayout(alerts, BoxLayout.PAGE_AXIS))
  getContentPane.add(infos)
  getContentPane.add(alerts)
  setVisible(true)

File: smart-hub/src/main/kotlin/io/github/tassiLuca/hubkt/adapters/ui/DashboardUI.kt
package io.github.tassiLuca.hubkt.adapters.ui

import java.awt.BorderLayout
import java.awt.Font
import java.awt.GridLayout
import javax.swing.BoxLayout
import javax.swing.JFrame
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTable
import javax.swing.WindowConstants
import javax.swing.table.DefaultTableModel

/** The dashboard Swing UI. */
class DashboardUI : JFrame() {

    /** The temperature label. */
    val temperatureLabel = JLabel("--")

    /** The heater label. */
    val heaterLabel = JLabel("--")

    /** The alerts model. */
    val alertsModel = DefaultTableModel(arrayOf<Any>("Alert Messages"), 0)

    private val biggerFont = Font("Arial", Font.BOLD, BIGGER_FONT_SIZE)

    init {
        defaultCloseOperation = WindowConstants.EXIT_ON_CLOSE
        title = "Smart hub dashboard"
        setSize(WIDTH, HEIGHT)
        setLocationRelativeTo(null)
        layout = GridLayout(1, 2)
        temperatureLabel.font = biggerFont
        heaterLabel.font = biggerFont
        val luminosityLabel = JLabel("--")
        luminosityLabel.font = biggerFont
        val infos = JPanel().apply {
            layout = BoxLayout(this, BoxLayout.PAGE_AXIS)
            add(JPanel())
            add(
                JPanel().apply {
                    add(JLabel("Current Temperature:"))
                    add(temperatureLabel)
                },
            )
            add(
                JPanel().apply {
                    add(JLabel("Heater Status:"))
                    add(heaterLabel)
                },
            )
            add(
                JPanel().apply {
                    add(JLabel("Current Luminosity:"))
                    add(luminosityLabel)
                },
            )
        }
        val alerts = JPanel(BorderLayout()).apply {
            add(JScrollPane(JTable(alertsModel)))
        }
        contentPane.add(infos)
        contentPane.add(alerts)
        isVisible = true
    }

    companion object {
        private const val WIDTH = 915
        private const val HEIGHT = 460
        private const val BIGGER_FONT_SIZE = 20
    }
}
